torch/__init__.py

```diff
@@ -745,6 +745,14 @@ def set_default_tensor_type(t):
     """
     if isinstance(t, str):
         t = _import_dotted_name(t)
+    import warnings
+    warnings.warn("torch.set_default_tensor_type() is deprecated as of PyTorch 2.1, "
+                  "please use torch.set_default_dtype() and torch.set_default_device() "
+                  "as alternatives.")
+    # torch.tensortype only alived in c++ side, so using type(torch.FloatTensor)
+    if not isinstance(t, type(torch.FloatTensor)):
```

Comment:
I'm pretty sure this isn't semantics preserving

Reply:
Yep, This check will be more generic for all tensor types. 
The content of the last discussion：
from Shan19900305：
PyTensorType_Check only checks object whether is in tensor_types, and it is assumed that the data stored on the torch._tensor_classes and tensor_types vector are consistent.
Now tensor_types are only used for CPU and GPU tensor types, which is a local static vector. But _tensor_classes is python side, which can be registered from other libraries for custom device tensor types. So from checked object based on local tensor_types vector to global torch._tensor_classes.
from ezyang：
Drop this bit from this PR, if you really want it let's litigate it on its own. But this is the same "I don't want to aid and abet more usage of the legacy constructor style"

And now split this bit to this PR.